ftl/common/builder.py

```diff
@@ -55,34 +65,108 @@ class JustApp(Base):
+                layer tarfile from context.')
+
+            tar = buf.getvalue()
+
+            with ftl_util.Timing("gzip_app_layer_tar"):
+                gzip_process = subprocess.Popen(
+                    ['gzip', '-f'],
+                    stdout=subprocess.PIPE,
+                    stdin=subprocess.PIPE,
+                    stderr=subprocess.PIPE)
+                gz = gzip_process.communicate(input=tar)[0]
+            self._img = tar_to_dockerimage.TarDockerImage(gz)
+
+
+class RuntimeBase(JustApp):
+    """Base is an abstract base class representing a container builder.
 
-    def BuildAppLayer(self):
-        """Override."""
-        buf = cStringIO.StringIO()
-        logging.info('Starting to generate app layer tarfile from context...')
-        with tarfile.open(fileobj=buf, mode='w') as out:
-            for name in self._ctx.ListFiles():
-                content = self._ctx.GetFile(name)
-                info = tarfile.TarInfo(os.path.join('app', name))
-                info.size = len(content)
-                out.addfile(info, fileobj=cStringIO.StringIO(content))
-        logging.info('Finished generating app layer tarfile from context.')
-
-        tar = buf.getvalue()
-        sha = 'sha256:' + hashlib.sha256(tar).hexdigest()
-
-        logging.info('Starting to gzip app layer tarfile...')
-        gzip_process = subprocess.Popen(
-            ['gzip', '-f'],
-            stdout=subprocess.PIPE,
-            stdin=subprocess.PIPE,
-            stderr=subprocess.PIPE)
-        gz = gzip_process.communicate(input=tar)[0]
-        logging.info('Finished gzipping tarfile.')
-        return gz, sha
+    It provides methods for generating a dependency layer and an application
+    layer.
+    """
+
+    def __init__(self, ctx, namespace, args, cache_version_str,
+                 descriptor_files):
+        super(RuntimeBase, self).__init__(ctx)
+        self._namespace = namespace
+        self._args = args
+        self._base_name = docker_name.Tag(self._args.base)
+        self._base_creds = docker_creds.DefaultKeychain.Resolve(
+            self._base_name)
+        self._target_image = docker_name.Tag(self._args.name)
+        self._target_creds = docker_creds.DefaultKeychain.Resolve(
+            self._target_image)
+        self._transport = transport_pool.Http(httplib2.Http, size=_THREADS)
+        self._cash = cache.Registry(
+            self._target_image.as_repository(),
+            self._target_creds,
+            self._transport,
+            cache_version=cache_version_str,
+            threads=_THREADS,
+            mount=[self._base_name])
+        self._base = docker_image.FromRegistry(
+            self._base_name, self._base_creds, self._transport)
+        self._base.__enter__()
+        self._descriptor_files = descriptor_files
+
+    def Build(self):
+        return
+
+    def AppendLayersIntoImage(self, lyr_imgs):
+        ftl_image = lyr_imgs[0]
+        for i in range(1, len(lyr_imgs)):
```

Comment:
nit: enumerate is a bit easier to follow than this style.

Reply:
done